File: prototype/rails/src/app/domain/shop/listing_search.rb
module Domain
  module Shop
    # What a storefront visitor asked to see: free text over the catalogue and a
    # medium to narrow it to.
    ListingSearch = Data.define(:term, :medium) do
      def self.from_input(term:, medium:)
        new(term: filled(term), medium: filled(medium))
      end

      def self.filled(value)
        text = value.to_s.strip

        text.empty? ? nil : text
      end
      private_class_method :filled

      def term?
        !term.nil?
      end

      def medium?
        !medium.nil?
      end

      # SQLite LIKE has no escape character unless the query names one, so a
      # wildcard the visitor typed is dropped rather than escaped.
      def like_pattern
        raise ArgumentError, "a search without a term has no pattern" unless term?

        "%#{term.tr('%_', '  ').squeeze(' ').strip}%"
      end
    end
  end
end

File: prototype/php/src/app/Domain/Shop/ListingSearch.php
<?php

declare(strict_types=1);

namespace App\Domain\Shop;

use DomainException;

/**
 * What a storefront visitor asked to see: free text over the catalogue and a
 * medium to narrow it to.
 */
final readonly class ListingSearch
{
    private function __construct(public ?string $term, public ?string $medium) {}

    public static function fromInput(?string $term, ?string $medium): self
    {
        return new self(self::filled($term), self::filled($medium));
    }

    public function hasTerm(): bool
    {
        return $this->term !== null;
    }

    public function hasMedium(): bool
    {
        return $this->medium !== null;
    }

    public function likePattern(): string
    {
        if ($this->term === null) {
            throw new DomainException('A search without a term has no pattern.');
        }

        // SQLite LIKE has no escape character unless the query names one, so a
        // wildcard a visitor typed is dropped rather than escaped.
        $literal = preg_replace('/\s+/', ' ', str_replace(['%', '_'], ' ', $this->term));

        return '%'.trim((string) $literal).'%';
    }

    private static function filled(?string $value): ?string
    {
        $trimmed = trim((string) $value);

        return $trimmed === '' ? null : $trimmed;
    }
}
